pkg/datasource/compat/wrapper.go

```diff
@@ -0,0 +1,155 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compat
+
+import (
+	"encoding/binary"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+)
+
+type EventWrapperBase struct {
+	ds                           datasource.DataSource
+	MntnsidAccessor              datasource.FieldAccessor
+	NetnsidAccessor              datasource.FieldAccessor
+	nodeAccessor                 datasource.FieldAccessor
+	namespaceAccessor            datasource.FieldAccessor
+	podnameAccessor              datasource.FieldAccessor
+	containernameAccessor        datasource.FieldAccessor
+	runtimenameAccessor          datasource.FieldAccessor
+	containeridAccessor          datasource.FieldAccessor
+	containerimagenameAccessor   datasource.FieldAccessor
+	containerimagedigestAccessor datasource.FieldAccessor
+}
+
+func WrapAccessors(source datasource.DataSource, mntnsidAccessor datasource.FieldAccessor, netnsidAccessor datasource.FieldAccessor) (*EventWrapperBase, error) {
+	ev := &EventWrapperBase{
+		MntnsidAccessor: mntnsidAccessor,
+		NetnsidAccessor: netnsidAccessor,
+	}
+
+	k8s, err := source.AddField("k8s", datasource.WithFlags(datasource.FieldFlagEmpty))
+	if err != nil {
+		return nil, err
+	}
+
+	runtime, err := source.AddField("runtime", datasource.WithFlags(datasource.FieldFlagEmpty))
+	if err != nil {
+		return nil, err
+	}
+
+	ev.nodeAccessor, err = k8s.AddSubField("node", datasource.WithTags("kubernetes"))
+	if err != nil {
+		return nil, err
+	}
+	ev.namespaceAccessor, err = k8s.AddSubField("namespace", datasource.WithTags("kubernetes"))
+	if err != nil {
+		return nil, err
+	}
+	ev.podnameAccessor, err = k8s.AddSubField("podName", datasource.WithTags("kubernetes"))
+	if err != nil {
+		return nil, err
+	}
+	ev.containernameAccessor, err = runtime.AddSubField("containerName")
+	if err != nil {
+		return nil, err
+	}
+	ev.runtimenameAccessor, err = runtime.AddSubField("runtime")
+	if err != nil {
+		return nil, err
+	}
+	ev.containeridAccessor, err = runtime.AddSubField("containerId")
+	if err != nil {
+		return nil, err
+	}
+	ev.containerimagenameAccessor, err = runtime.AddSubField("containerImageName")
+	if err != nil {
+		return nil, err
+	}
+	ev.containerimagedigestAccessor, err = runtime.AddSubField("containerImageDigest")
+	if err != nil {
+		return nil, err
+	}
+	return ev, nil
+}
+
+type EventWrapper struct {
+	*EventWrapperBase
+	Data datasource.Data
+}
+
+func (ev *EventWrapper) GetMountNSID() uint64 {
+	d := ev.MntnsidAccessor.Get(ev.Data)
+	switch len(d) {
+	case 4:
+		return uint64(binary.LittleEndian.Uint32(d))
+	case 8:
+		return binary.LittleEndian.Uint64(d)
```

Comment:
No, we don't - but we need a clean way to handle data here. I've experimented a bit with it, but I don't want us to have to use the unsafe package in operators (let alone worry about alignment), so I'll probably add the `NativeEndian` thing that cilium also uses internally (probably for the same (and hopefully good) reason).

Same applies to the `formatter` operator.

Reply:
A new `ByteOrder()` function has been added to the DataSource interface - depending on how the DataSource is created (locally or using gRPC), the correct ByteOrder is assigned. So `ds.ByteOrder().Uint32(d)` will always use the native endianess of the host that the data originated from. I think that's a good compromise.

Reply:
> > In the wrapper.go, we're enriching data that has locally been emitted. So for that case it's true that we don't need to handle endianness, we can assume a cast would suffice. However, I think the way to cast in Go is rather ugly (and IMHO in operators we shouldn't use ugly things), so I opted for clean helpers.
> 
> I do not like `golang` casting too, as I am more into `C` way of doing it. But we should stick to what the language offers, particularly as built-ins are, I guess, better handled with regard to performance than function calls.

The binary package is what is encouraged to be used when dealing with binary data (which is what we're doing here), so I think it's not a bad fit. It might be a tad slower (if you look at the `encoding/binary` packages' source you'll find it is really simple and probably written in a way that the compiler can optimize further) compared to actual casting, but it makes it much easier to read IMHO.

> > In addition to that, we have to take care of endianness when crossing machine boundaries, so this also adds support for that. Depending on the system that created the DataSource, the endianness is set. So even if the server (in a client/server scenario) emits data using LE, an operator on the client using BE could use the ds.ByteOrder() to be able to extract the correct values. On the client side, DataSources could also be created; in that case, the DataSource would have BE as default.
> 
> Why would `DataSource` created by the client be big endian rather than the client endianess?

A DataSource would always have the endianness of the host that is creating the DataSource - in my example the client just happens to be big endian. Since the ByteOrder would automatically be swapped in that case, you can simply use the same functions and always get the proper values for the current host.

> 
> > This theoretically allows complex setups like client -> server -> other server, where each system runs operators and has a different endianness without thinking about endianness at all. (If that example is wild, think of a map/reduce situation that could aggregate stuff on the intermediary server.)
> 
> I may be too pragmatic, but we should focus on situation which are very like to occur. Out of the blue, I really do not see any real life situation where someone would use Inspektor Gadget like this. Moreover, we should really focus on simplicity and not offering ways of doing complex stuff. If users really want to suffer doing complex stuff, this is up to them but clearly not our problem.

The scenario is not that unlikely IMHO - however, it was just an example to emphasize that even those complex cases would automatically be handled correctly using the current approach. I think it's almost free lunch :)